### ISSUE TITLE
[Returned for Revisions] SDL 0303 - SDL Audio I/O Manager

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#1013]
Hello SDL community,

The review of "SDL 0303 - SDL Audio I/O Manager" begins now and runs through September 22, 2020. Previous reviews of this proposal took place May 19 - 26, 2020 and June 17 - July 14, 2020. The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0303-audio-io-manager.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/1013

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
1\. This proposal is in conflict with the accepted proposal [SDL-0295](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0295-screen-manager-audio-pass-thru.md). I would recommend that this proposal be rejected in favor of a revisions proposal to that one (or a new proposal) that adds some of the features from this proposal (namely, pausing audio streaming output when input is requested and audio level normalization) to the structure of that proposal.

2\. I don't believe that having the entirety of the code be in the proposal is a good idea. The maintainers must effectively attempt to integrate the code and provide a code review in the proposal comments for the author to update. Such a process would take a _very_ long time and is, in my view, a waste of everybody's time.

3\. I don't see how we could accept this proposal without corresponding Java and JavaScript impact. Adding features to iOS only that can and should be on all app libraries undos a lot of work that the PM has gone through to make the libraries more aligned.

Given those recommendations, I will hold off reviewing the rest of the proposal until the SDLC has decided how to proceed with this proposal.

**Comment by theresalech** [comment]
The Steering Committee voted to defer this proposal until the author is able to respond to the comments provided on the review issue.  Once the author is able to respond, he can email SDLC@smartdevicelink.com, and the proposal can be brought back under Steering Committee review.

**Comment by theresalech** [comment]
The author has advised that he is able to respond to the comments provided on the issue.  This proposal is ready to be brought back into review, pending Steering Committee vote.

**Comment by kshala-ford** [comment]
1\. It's disappointing to see that nothing of our internal discussions is mentioned in your comment. We discussed how your suggested APIs can go hand in hand with the I/O manager. Considering the history and importance of this I/O manager I would have appreciated the PM to at least inform the SDLC members about the existence and other conversations that have happened. I couldn't find anything in the comments of 0295...  

The `SDLMicrophoneInputView` provides much less features than the I/O manager. As the name already indicates it belongs to data input. I don't know how to add output capabilities without an overhaul of this class. 

My suggestion is to add the I/O manager to the SDL libraries and have the MicrophoneInputView or actually the screen manager use the I/O manager as a submanager. This way all applications especially those with inability of audio output can use the input view. Other applications like mobile navigation apps can also use the input view or directly use the I/O manager.

2\. The code is already in use in production therefore why not add the code to the proposal? Why holding back information from the SDLC? With this proposal I am not holding back any available information. If you can explain why I should do that please feel free to comment. 

The proposal acceptance shouldn't include the manager implementation itself though I highly recommend the implementation to be the base for the pull request. However the public API and every feature that this manager serves should be voted on. You will find the public API in the header files. I was hesitant to create a copy in the proposal file to avoid keeping header files and proposal content in sync. The features and workarounds are all part of the proposal and described to my best knowledge.

3\. As mentioned the code already exists. Unfortunately only for iOS at the moment. We can for sure add the other platforms and describe their public APIs. As already mentioned all features that the manager serves are already described in the proposal and apply to all platforms.

**Comment by joeljfischer** [comment]
I apologize for not responding earlier, I was out most of the review cycle.

1\. I'm sorry that I don't recall the details of that discussion. It was many months ago and there is no longer a record of it. I thought that we had agreed to have the `MicrophoneInputView` proposal go through as is and to have this proposal be modified to act as a supporting proposal that adds additional features to that proposal. Because this proposal exists unmodified, I assume that one of us was wrong in our current impressions of that past conversation.

1a. 
> The SDLMicrophoneInputView provides much less features than the I/O manager. As the name already indicates it belongs to data input. I don't know how to add output capabilities without an overhaul of this class.

I don't believe that output capabilities are necessary for this class. The existing `SDLAudioStreamManager` and `SDLStreamingMediaManager` work fine as-is and it's better to not mix these two. Only a small percentage of apps need audio output via SDL. It may in fact confuse developers to have audio output outside of the streaming media module.

Correct me if I'm wrong, but there are effectively two major features of this proposal compared with SDL-0295 in my understanding of this proposal: (1) the ability to *pause* the input audio when audio stream output begins and to resume it when it ends is what is important. I believe that this can be accomplished using `NSNotification`s without tight coupling. It's not something that I want to do often, but in this case I think it makes sense. (2) The other major feature I see, to have the input audio be modified for volume, could easily be integrated as a feature of the `MicrophoneInputView`.

> My suggestion is to add the I/O manager to the SDL libraries and have the `MicrophoneInputView` or actually the screen manager use the I/O manager as a submanager. This way all applications especially those with inability of audio output can use the input view. Other applications like mobile navigation apps can also use the input view or directly use the I/O manager.

I don't think that this is useful. Now we have a very awkward coupling of sub-managers where some developers use one, some use another, and some use a third, and they all are mixed together with dependencies. That's not a place that I'd like to see our design go. As I noted above, I think that adding features and loose coupling where necessary to the `MicrophoneInputView` and `AudioStreamManager` is a better way to go.

2\. 

> The code is already in use in production therefore why not add the code to the proposal? Why holding back information from the SDLC? With this proposal I am not holding back any available information. If you can explain why I should do that please feel free to comment.

I understand providing sample code, however, I think that now we're mixing up our issues. When a full suite of code is provided it becomes difficult to know for certain what code is intended for review and what code is intended as a sample. In my mind (this is not the stated process of the SDLC, just my opinion, which has evolved over time), what is necessary for me to review the proposal is: (1) the exact public APIs that will be used, (2) sample code that is necessary for (a) proving that a feature is possible, or (b) the approach that will be taken if multiple approaches are possible.

> The proposal acceptance shouldn't include the manager implementation itself though I highly recommend the implementation to be the base for the pull request.

Given my comments on (1) above, I believe that this proposal should be restructured to fit with the rest of the SDL app library and not the other way around. This code was made for one app's needs, outside of SDLC approval, and to me, we need to make sure that we approach the problem with fresh eyes and make it a good interface for all developers. I think that the changes I described there will do that.

> However the public API and every feature that this manager serves should be voted on. You will find the public API in the header files. I was hesitant to create a copy in the proposal file to avoid keeping header files and proposal content in sync. The features and workarounds are all part of the proposal and described to my best knowledge.

The code is sample code, but the APIs are exactly what should be implemented? That needs to be made clear. I think that the exact APIs to be added need to be in the proposal text itself to help clarify. Additionally, as I noted, because of the approach of SDL-0295, I don't think that this class needs to exist at all and can be used as a no-API addition to the `MicrophoneInputView`.

3\. 👍 Because we don't want to diverge on our features as much as possible, I think that it's necessary to apply this to Java Suite and JavaScript Suite. You can always leave it up to the maintainers' discretion to ensure that APIs are in alignment outside of major changes.

**Comment by theresalech** [comment]
The Steering Committee voted to keep this proposal in review, to allow the author time to respond to the last comment from the Project Maintainer.

**Comment by kshala-ford** [comment]
1\. I think we have a different memory to that conversation but as you say there are no records anymore. Back to your original point of rejecting this proposal I want to say that I fully disagree. This proposal adds audio input and output as a whole while your proposal added a view based screen manager API for audio input only.

1a\. My message was to say that we shouldn't add output capabilities to the `SDLMicrophoneInputView`. As the name says it's related to input audio as a view.

The `SDLAudioStreamManager` does not work fine as-is because it breaks with the audio pass thru RPC. You can't play audio while listening on the micrphone. Both SDL-0295 *and* the audio stream manager must align with each other. My suggestion is to deprecate the old stream manager as a whole and replace it with the IO manager for many different reasons. Not only to synchronize with audio input but also to address a race condition of an empty stream although data is still in process.

(1) The ability to pause the input stream and synchronize with audio output is the key feature of this proposal. Using `NSNotification` to synchronize with your input view is a very bad idea to be honest. The input and output synchronization is bidirectional and isn't resolved in a single notification in one direction. The output must not start until the input as properly closed resulting in multiple messages in both directions. Same goes for the input stream which must not start as long as audio is left in the output stream. Adding the known issues of Core 4.2 as I described in the proposal I still suggest to add one class which encapsulates both audio stream directions.

(2) Sure this can be done but why? The audio input improvements should belong to where the audio input is managed. Audio input and audio output are so dependent from each other let alone this is within the context of audio streaming. I don't see any reason to split the logic into a manager and a view. Again. the `MicrophoneInputView` should be calling the manager for action.

2\.

To resolve your concerns both header files are part of the proposal and I am happy to copy them into the .md file to make that clear. I don't want to have the .m file be part of the proposal acceptance but I highly recommend it should be the base for implementing this proposal.

TL;DR I think we don't get to an agreement here. The PM are defending the `MicrophoneInputView` and extend `AudioStreamManager` which is a very difficult task and requires a rather poor API design. As the PM states `It's not something that I want to do often` I don't see why we should do it now if it's not necessary.

I suggest to have audio related functionality being coupled in one context which is the Audio I/O manager and the `MicrophoneInputView` should leverage from that manager.

**Comment by joeljfischer** [comment]
1\. To be clear, when I say, "reject" this proposal, I mean to turn it into a revisions proposal for the microphone input manager.

1a\. I'm in agreement that we should not add output capabilities to the `Microphone Input View` (nor the `MicrophoneInputManager`), not because it's a view, but because it's scoped for microphone input. But features that allow it to _work_ when SDL is outputting audio are fully in scope.

1.a.1\.

> Using NSNotification to synchronize with your input view is a very bad idea to be honest. The input and output synchronization is bidirectional and isn't resolved in a single notification in one direction. The output must not start until the input as properly closed resulting in multiple messages in both directions.

From what I can see from your description, there's three or four (input start/stop, output start/stop). I agree that it's not ideal, but it's not especially unwieldy either. We've done similar things in the past (e.g. `SDLCarWindow` and the lock screen, which uses four notifications). 

The alternative is to more tightly couple the managers and allow them to message each other by using a protocol system (each conforms to a protocol the other understands). Again, this isn't ideal, but we've done similar things in the past (e.g. passing `SDLConnectionManagerType` to sub-managers to allow sending RPCs without passing the full `SDLLifecycleManager` API).

Either of these solutions, to me, is much better than the solution proposed here of deprecating audio output from the streaming media module, moving it out of the streaming media module in an object combining that and audio input (which breaks the separation of scopes unnecessarily).

> Adding the known issues of Core 4.2 as I described in the proposal I still suggest to add one class which encapsulates both audio stream directions.

I don't see a reference to Core 4.2 in your proposal, but if there's a bug (perhaps the "race condition" in the proposal), this proposal is not the forum to discuss and fix it. You should open an issue on the relevant Github repository and discuss it there. The fix you propose in your manager seems like a significant issue that should be considered carefully and not pushed in with a broader proposal for API refactors/changes. We would need to make sure that everything works as intended with all SDLC HUs and apps. If the scope is greater than a bugfix (which it seems it might be, given the behavior changes noted), it should ideally be made in a separate proposal, given that the fix seems like it could easily be transferred to the existing `SDLAudioStreamManager`.

1.a.2\. The "view" is not the only part of that proposal. There's a Screen Manager sub-manager (`SDLMicrophoneInputManager`) that manages the audio input. That is the place where we would add the audio normalization.

2\. Thank you, that would help in understanding exactly what is being proposed and required for implementation.

> TL;DR I think we don't get to an agreement here. The PM are defending the MicrophoneInputView and extend AudioStreamManager which is a very difficult task and requires a rather poor API design. As the PM states It's not something that I want to do often I don't see why we should do it now if it's not necessary.

I think it's very early to say that there's no room for compromise or agreement unless you're stating that you have no intention of attempting to understand the PM's viewpoints or changing your mind.

I do believe that the `MicrophoneInputManager` could easily scale to handle the feature of audio normalization, and I believe that mechanisms exist to handle managing audio output / input in the existing managers without re-architecting the whole thing in a "rather poor API design". It's unnecessary, and this proposal could be a simple extension of the existing and accepted solutions with a fraction of the work and API changes. I understand that you want this proposal accepted as-is because you already implemented these changes into an app outside of SDLC approval and you don't want that app to have to make changes to accommodate a different design, but that's not how changes should find their way into the SDL libraries.

**Comment by kshala-ford** [comment]
Instead of replying with yet another large comment I would like to have a workshop with the PM to resolve the conflicts between SDL-0295 and this proposal.

**Comment by theresalech** [comment]
We did not have a quorum present during the 2020-07-07 Steering Committee meeting, so voting on this proposal could not take place.  This proposal will remain in review until 2020-07-14.

**Comment by theresalech** [comment]
The Steering Committee voted to defer this proposal in favor of a separate workshop to discuss open items, especially as they relate to the accepted proposal SDL 0295.

**Comment by theresalech** [comment]
The workshop regarding this proposal has taken place, and the following were agreed upon by the author and PM as requested revisions to the proposal:
1\. Add following statement to the proposal (note this is just adding it to the proposal, not accepting this statement):
> SDL does not include requirements to head units to support `PerformAudioPassThru` and Audio services to run at the same time.
> We can't expect and we cannot guarantee that all SDL enabled head units will support both to work at the same time. The current SDL library don't prevent an app from executing both.
> Therefore receiving audio data with `PerformAudioPassThru` and sending audio data with the Audio Service must be synchronized.

2\. Create state machine diagrams to help understanding the challenges of synchronizing input stream and output stream.

3\. Rework the sub-section of workarounds so that all workarounds made are more visible and better to understand.

The proposal is now ready to be brought back under Steering Committee review, so that the committee can review these requested revisions and vote on returning the proposal for such.

**Comment by theresalech** [comment]
It was requested that this proposal not be brought back into review until the author is available to respond to any comments provided on the review issue. The author has advised that he is now available to respond, so the Steering Committee can vote to bring this back into review when time allows.

**Comment by theresalech** [comment]
The Steering Committee voted to bring this proposal back into review.

Steering Committee representatives should review the requested revisions outlined in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/1013#issuecomment-682101300) and plan to vote on this proposal during the September 22, 2020 Steering Committee meeting.

**Comment by theresalech** [comment]
The Steering Committee voted to return this proposal for revisions. The author will revise the proposal to incorporate the revisions listed in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/1013#issuecomment-682101300).

**Comment by theresalech** [comment]
Closing as inactive. The issue will remain in a `returned for revisions` state. The author should notify the Project Maintainer when a revisions PR has been submitted.
